### DISCUSSION TITLE
Freeing Shapes on Body : Emscripten --memoryprofiler

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#99] [Q&A] [answered]
I know that general cleanup and memory handling was discussed in https://github.com/jrouwe/JoltPhysics.js/discussions/63, but wanted to isolate this to just my tests on memory profiling and Creation/Destruction of Shapes associated with Bodies.

I found the emscripten memory profiler, so styled it to load on all Jolt demos I was running for checking my leaks.

I was checking through some of my test cases for leaks, since I planned on doing some game-level loading in another project that needs to do cleanup between levels.
https://github.com/PhoenixIllusion/JoltPhysics.js/tree/feature/leak-shape-debug

I found some leaks for Vec3's on the Character Virtual, but after that my main focus was on creation/destruction since that will be a key part of level-loading.

I found on `/add_remove_bodies.html` that if the `Shape` is created inside the `generateObject()` loop it leaks noticeably over time, but if you move just that outside the loop and keep the rest of the `pos`, `size`, `rot` creation/destruction it stabilizes around at-most 1026 reference-counts.

I moved those common variables outside the loop so I could isolate the issue to just `shape` ref issue, and was unable to figure out what logic I could use for actually freeing the `shape` and preventing the leak. Moving the shape-creation outside the `generateObject()` method, even after running for minutes it stays at a fixed Ref-count and no malloc-growth.
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/70cabe31-a0a3-437c-9465-5ec3626aaec0)

I found that if BoxShape is not used in `BodyCreationSettings`, just randomly calling `new`/`destroy` in the loop properly cleans it up, and calling `AddRef`/`Release` in the loop cleans it up, but as soon as it was passed into the `BodyCreationSettings` it will start leaking or not be properly cleaned up by the `Body` destruction
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/98a6fa61-aee2-400b-96ff-ec4ff93313de)

I tried keeping a reference myself, using `AddRef` so `Body` wouldn't auto-destroy and trying to manually free it with `destroy`, or keeping 1x extra `AddRef` myself and calling `Release` post-body-destruction, but nothing I did really seemed to free it and stabilize the living-Ref-count in the profiler.

So far, from my tests, I've found the only leaks are new shapes passed to the body. If I restrict my game Levels to a pool of finite Shapes such that they all can be in memory at once, I can make as many bodies as I want and clean up everything 100%. This is kind of workable for the near future, but ideally I can find some way to actually destroy shapes and do clean rebuilds of Levels.

----

Edit:
Looking through this, it appears to actually be leaking Bodies? 

If I skip the `AddBody`/`RemoveBody` from the physics system and just purely use `bodyInterface.CreateBody` and `bodyInterface.DestroyBody` in the same `generateObject` flow, skipping the actual physics simulation and ThreeJS, it still has the leaks.

If I create a pool of 100 shapes, and loop through and just assign unused ones to 100 spawned boxes, and do the same logic, I can freely create/destroy 100 bodies, either with `CreateBody`/`DestroyBody` or also in addition to `AddBody`/`RemoveBody`

**Comment by jrouwe** [comment]
There is something funny going on indeed. If I however change the sample to this:

```
function generateObject() {
	// Position and rotate body
	let pos = new Jolt.Vec3((Math.random() - 0.5) * 25, 15, (Math.random() - 0.5) * 25);
	let rot = getRandomQuat();

	// Create physics body
	let size = new Jolt.Vec3(0.5, 0.5, 0.5);
	let shape = new Jolt.BoxShape(size, 0.05, null);
	Jolt.destroy(size);
	let creationSettings = new Jolt.BodyCreationSettings(shape, pos, rot, Jolt.EMotionType_Dynamic, LAYER_MOVING);
	Jolt.destroy(pos);
	creationSettings.mRestitution = 0.5;
	let body = bodyInterface.CreateBody(creationSettings);
	Jolt.destroy(creationSettings);

	addToScene(body, 0xff0000);
	removeFromScene(dynamicObjects[dynamicObjects.length - 1]);
}

function onUpdate(time, deltaTime) {
	// Check if its time to spawn a new object
	if (time > timeNextSpawn) {
		for (let i = 0; i < 100; ++i)
			generateObject();
		timeNextSpawn = time + objectTimePeriod;
	}
}
```

(revert back to original with change: add and remove the same body immediately and do that 100x per frame)

Then I don't see any leaks.

**Reply by PhoenixIllusion** [reply]
Yes, I saw the leak only after you start adding and removing using a queue of the Bodies, most notably about 10-20 items in length. Immediate creation and destruction of the bodies doesn't seems to leak.

The fact that, in my tests, it is growing so slow means that it isn't every Body causing the issue, there is some more random cause. The code is doing the add/remove at 100 times a second or more, but the leak seems to be just 10-20 per second.

Random behavior makes me think shared addresses that are reused over cycles may have some involvement. I was going to try reproducing in the JoltPhysics CPP layer, skipping JS, since this doesn't seem concretely JS related off hand.

**Comment by jrouwe** [comment, marked as answer]
My conclusion is that the memoryprofiler is not tracking allocations correctly. If I add the following code to Shape.h:

```
	inline static int test_num_shapes = 0;

	/// Constructor
									Shape(EShapeType inType, EShapeSubType inSubType) : mShapeType(inType), mShapeSubType(inSubType) { ++test_num_shapes; }
									Shape(EShapeType inType, EShapeSubType inSubType, const ShapeSettings &inSettings, [[maybe_unused]] ShapeResult &outResult) : mUserData(inSettings.mUserData), mShapeType(inType), mShapeSubType(inSubType) { ++test_num_shapes; }

	/// Destructor
	virtual							~Shape()
	{
		--test_num_shapes;
		std::cout << "Num shapes: " << test_num_shapes << std::endl;
	}
```

And the following to Body.h:

```
	inline static int test_num_bodies = 0;

	/// Default constructor
							Body()
	{
		++test_num_bodies;
	}

	/// Destructor
							~Body()
	{
		JPH_ASSERT(mMotionProperties == nullptr);
		--test_num_bodies;
		std::cout << "Num bodies: " << test_num_bodies << std::endl;
	}
```

Then it just prints every frame:

```
Num bodies: 200
Num shapes: 201
```

**Reply by PhoenixIllusion** [reply]
The emscripten available memory appears to still be decreasing per untracked body, in line with the size per body it says is leaked. The Shapes don't seem to really be the memory hog, but the KB's per body has consumed more than a MB after running for a while. The tracker itself just appears to track malloc() and free(), and keeps track of any value from malloc() that doesn't get passed to free(). The log of allocations in the profiler tool has records of which pointers have met those conditions of malloc with no free.

I'm going to try a stripped down sample and see if I can get it to the case of just BodyManager and "in this case, the 100th iteration tries to destroy and the free doesn't report the pointer"

**Reply by PhoenixIllusion** [reply]
I guess the answer is your 'not tracking allocations correctly'. The memory profiler's modifications, hooking into the `malloc`, are messing up when tracing the custom alignment code of Jolt. The addresses being allocated do not match the returned pointers for `Body`. The addresses of the Body allocated and free'd are sometimes off by about +/- 24 or something. I think the memoryProfiler is not actually checking the true Free space and may be just checking it's internal totals. I may look into how to fix that to make the profiler work as expected but for now will fall back to raw FreeMem checks.
- https://github.com/emscripten-core/emscripten/blob/main/src/memoryprofiler.js : code says if the free was done on a alignment-shifted, the non-exact match will cause a skip of the `totalMemoryAllocated` change since it didn't see the original value, thus appear as a (non real) leak.

Known Issue: https://github.com/emscripten-core/emscripten/issues/6637

The issue seemed to go away for now when I removed the profiler code and stuck with a raw 'getFreeMemory' code added to the Jolt Interface - https://github.com/emscripten-core/emscripten/issues/6637#issuecomment-394791802

Running the regular Add/Remove code and just manually keeping track of the 'FreeMemory' value appears to remove any leak that was previously there.

**Reply by jrouwe** [reply]
Interesting! Yes that explains it.

For your test you could change:

```
#elif defined(JPH_PLATFORM_WASM)
	// WebAssembly CPU architecture
	#define JPH_CPU_WASM
	#define JPH_CPU_ADDRESS_BITS 32
	#define JPH_VECTOR_ALIGNMENT 16
	#define JPH_DVECTOR_ALIGNMENT 32
	#define JPH_DISABLE_CUSTOM_ALLOCATOR
```

to

```
#elif defined(JPH_PLATFORM_WASM)
	// WebAssembly CPU architecture
	#define JPH_CPU_WASM
	#define JPH_CPU_ADDRESS_BITS 32
	#define JPH_VECTOR_ALIGNMENT 8
	#define JPH_DVECTOR_ALIGNMENT 8
	#define JPH_DISABLE_CUSTOM_ALLOCATOR
```

in Jolt/Core/Core.h (located in Build\\Debug\\_deps\\joltphysics-src\\Jolt\\Core\\Core.h). At this moment SIMD is off for the WASM build (due to Apple only recently getting support for it), so it shouldn't matter.

There are a few other aligned allocs in Jolt, so it will not cover 100% of the cases, but in this particular test it does fix the reported leaks.

**Comment by PhoenixIllusion** [comment]
As a heads-up to anyone else who wants to fully clean up Shapes, the following actually creates 2x RefCounts, which I was tracking down as a leak for hours.

The MeshFloor level would never give up 16k of data every `destroy` of the level. For some reason it didn't really 'leak', since re-creating the level and re-destroying over-and-over seemed to never grow the allocated no matter how many PhysicsSystems I rebuilt:

```
  const shapeSettings  = new Jolt.CylinderShapeSettings(0.5 * extents.y, 0.5 * extents.x);
  const result = shapeSettings.Create();
  const shape = result.Get(); // Shape starts with RefCount of 2
  const bodySettings = new Jolt.BodyCreationSettings(shape, pos, rot, Jolt.EMotionType_Dynamic, LAYER_MOVING))

  Jolt.destroy(shapeSettings); //Ref 1
  Jolt.destroy(result); //The mCacheResult holds Ref 2, so DestroyBody doesn't release 'shape'
```

I can now loop my BabylonJS plugin scene create/destroy cycle every 0.5 seconds through all my demos. I still have a 32 byte kind-of-leak somewhere that seem static so doesn't grow each cycle. Lost reusable Vec3 or something...

```
        import scene0 from './src/scene/2d_funnel';
        import scene1 from './src/scene/conveyor_belt';
        import scene2 from './src/scene/falling_shapes';
        import scene3 from './src/scene/friction';
        import scene4 from './src/scene/motor';
        import scene5 from './src/scene/sphere_bounce';
        import scene6 from './src/scene/vehicle_motorcycle';
        import scene7 from './src/scene/vehicle_wheeled';
        import scene8 from './src/scene/virtual_character';

        new App([
          scene0,scene1,scene2,scene3,
          scene4,scene5,scene6,scene7,scene8
        ]);
        
        // -----
        
        setInterval(async () => {
            clearMaterials();
            scene.dispose();
            
            console.log('Mem: ', scene_index, startMem - getFreeMemory());
            scene = await createNextScene();
          });
```

**Reply by PhoenixIllusion** [reply]
Not sure why the free of settings doesn't trigger the free of the cache and RefCount, unless the assignment during the chain does something? The Create().Get()

Just free'ing the ShapeSettings still leaves a RefCount of 1

Not sure if it is more proper to just always Release() post Create since I don't want to double-free the Cache, though a process that requires a user know they need to always `Release` just in this case due to a secret Ref is weird.

**Reply by PhoenixIllusion** [reply]
@jrouwe - Discovered my 32 byte issue was JS library's Jolt Interface destructor.

JoltInterface Constructor calls 'Register Types' which creates the default Physics Material (32 bytes)
https://github.com/jrouwe/JoltPhysics/blob/a6c9f8808f78c760ed0bed6f2e6c6129e66e9203/Jolt/RegisterTypes.cpp#L172
```
        PhysicsMaterial::sDefault = new PhysicsMaterialSimple("Default", Color::sGrey);
 ```
        
JoltInterface Destructor does not call Unregister Types, just destroys the `Factory::sInstance`

Since this is a static variables, and is null-pointer checked, it's not critical, but does make the following resolve to `0` instead of `32`

```
const memStart = getFreeMemory();
initPhysics();
Jolt.destroy(jolt);

console.log('Mem: ', memStart - getFreeMemory())
```
If Unregister was not done, it would just mean you had a floating 32 byte PhysicsMaterial forever in RAM, likely never causing any harm.

Only other notable personal leak to me was that VirtualCharacter's supportingShape is a `Plane` assigned as Value (non-Ref, non-Ref<Shape>, non-Pointer) so we have to do this:
```
    const plane = new Jolt.Plane(Jolt.Vec3.sAxisY(), -1);
    settings.mSupportingVolume = plane;
    Jolt.destroy(plane);
```

With the interface test, looped create/destroy of the different scenes (motors, characters, vehicles, etc) in my tests is all resolving to 0-bytes per test-index:
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/6963991f-a686-4bc4-99e9-b378435b4793)

**Reply by jrouwe** [reply]
Very nice! I've added the call to `UnregisterTypes()`.

~~What does your implementation of ```getFreeMemory()``` look like? I'd like to add that to ```proper_cleanup.html``` to ensure that it indeed cleaned up properly.~~ Nevermind I'm blind.

**Reply by PhoenixIllusion** [reply]
I've been baking in the version here:
https://github.com/emscripten-core/emscripten/blob/main/test/core/test_mallinfo.c
```
size_t getTotalMemory() {
  return (size_t)EM_ASM_PTR(return HEAP8.length);
}

size_t getFreeMemory() {
  struct mallinfo i = mallinfo();
  uintptr_t totalMemory = getTotalMemory();
  uintptr_t dynamicTop = (uintptr_t)sbrk(0);
  return totalMemory - dynamicTop + i.fordblks;
}
```

With regard to how new ShapeSetting-created Shapes start out with 2x Refs, what is the proper handling of the Free-ing of a new Shape, created via ShapeSettings?
Should we be calling `destroy` on the ShapeResult, or just always call `Release` on the Shape, regarding the above code:
```
  const shapeSettings  = new Jolt.CylinderShapeSettings(0.5 * extents.y, 0.5 * extents.x);
  const result = shapeSettings.Create();
  const shape = result.Get(); // Shape starts with RefCount of 2
  const bodySettings = new Jolt.BodyCreationSettings(shape, pos, rot, Jolt.EMotionType_Dynamic, LAYER_MOVING))

  Jolt.destroy(shapeSettings); //Ref 1
  Jolt.destroy(result); //The mCacheResult holds Ref 2, so DestroyBody doesn't release 'shape'
```

**Reply by jrouwe** [reply]
> 've been baking in the version here

Thanks!

> Should we be calling destroy on the ShapeResult, or just always call Release on the Shape, regarding the above code

No the problem is that ShapeResult has a reference to the shape, but it is a static variable inside the C++ Create() function, so it will not be released until the next time you call Create() on another ShapeSettings object. I've exposed a Clear() function to be able to release it:

https://github.com/jrouwe/JoltPhysics.js/commit/3e709033101fa4b05ee6927f2c65f6ce53371048

**Reply by jrouwe** [reply]
(btw that static variable is created by the webidl binder so I have no control over it)

**Reply by PhoenixIllusion** [reply]
My understanding of that is just that we assume the static's are allocated compile-time of Jolt, so that mem is locked up anyway and just 'the cost of the library'. Those addresses appear marked out and allocated in the WASM as constant addresses not requiring `malloc`.

That static storage being allocated once per ShapeSetting type explains why after running "Falling Shapes" the memory spikes so large and stays, given that test is kind of extensive in using almost every possible shape-setting, ensure we fill almost every static-variable mResultCache we possibly could, even after destroying Jolt and starting up a new physics simulation on a different use-case.

The concept of 'Clear()' seems more explicit than `Release()` and handles 1x more case I didn't run into.
In success cases, it appears the `Clear()` is the same as just calling 'Shape.Release()', except also handles possible leaks of freshly allocated mError `String`s, should the Result have been a failure.

`ShapeResult` is `Result<Ref<Shape>>`
`Result.Clear()` will call `~Ref<Shape>`
```
  case EState::Valid:
	mResult.~Type();
	break;

case EState::Error:
	mError.~String();
	break;
```
which I believe is "cleanup" but not `delete/free`

`mResult.~Ref()` which called `Ref<Shape>.Release()`
```
	inline  ~Ref() { Release(); }
```

**Reply by jrouwe** [reply]
What happens is that webidl_binder.py generates the following code:

```
Shape::ShapeResult* EMSCRIPTEN_KEEPALIVE emscripten_bind_SphereShapeSettings_Create_0(SphereShapeSettings* self) {
  static Shape::ShapeResult temp;
  return (temp = self->Create(), &temp);
}
```

(plus variants for each of the shape types)

So indeed at library creation time we have preallocated the ShapeResults for all different shape types. If you call:

```
let shapeSettings = new Jolt.SphereShapeSettings(1);
let result = shapeSettings.Create();
let shape = result.Get();
shape.Release();
```

Then the shape is indeed destroyed but the problem is that `result` now has a dangling pointer to `shape`, so next time you create another SphereShape through a SphereShapeSettings it will internally call `Release` and will attempt to free already freed memory. 

The `Clear` function internally indeed calls `Release` and it also resets the internal pointer so that the ShapeResult object is ready for reuse. I think you always need to call `Release` the same amount of times that you have called `AddRef` or else you may mess up Jolt's internal bookkeeping.

**Comment by jrouwe** [comment]
I liked your memory profiler integration, so now you can turn it on through an option, see: https://github.com/jrouwe/JoltPhysics.js/commit/1d4070427cad011ef27b8d0bc91a32daf099955d.

I also exposed the function to get memory usage in https://github.com/jrouwe/JoltPhysics.js/commit/22bdd1ccda270544dba976a2e3976779f1974804.
